third_party/move/move-binary-format/src/compatibility.rs

```diff
@@ -214,46 +271,169 @@ impl Compatibility {
+fn struct_abilities_compatible(old_abilities: AbilitySet, new_abilities: AbilitySet) -> bool {
     old_abilities.is_subset(new_abilities)
 }
 
 // When upgrading, the new type parameters must be the same length, and the new type parameter
 // constraints must be compatible
-fn fun_type_parameters_compatibile(
+fn fun_type_parameters_compatible(
     old_type_parameters: &[AbilitySet],
     new_type_parameters: &[AbilitySet],
 ) -> bool {
     old_type_parameters.len() == new_type_parameters.len()
         && old_type_parameters.iter().zip(new_type_parameters).all(
             |(old_type_parameter_constraint, new_type_parameter_constraint)| {
-                type_parameter_constraints_compatibile(
+                type_parameter_constraints_compatible(
                     *old_type_parameter_constraint,
                     *new_type_parameter_constraint,
                 )
             },
         )
 }
 
-fn struct_type_parameters_compatibile(
+fn struct_type_parameters_compatible(
     old_type_parameters: &[StructTypeParameter],
     new_type_parameters: &[StructTypeParameter],
 ) -> bool {
     old_type_parameters.len() == new_type_parameters.len()
         && old_type_parameters.iter().zip(new_type_parameters).all(
             |(old_type_parameter, new_type_parameter)| {
-                type_parameter_phantom_decl_compatibile(old_type_parameter, new_type_parameter)
-                    && type_parameter_constraints_compatibile(
+                type_parameter_phantom_decl_compatible(old_type_parameter, new_type_parameter)
+                    && type_parameter_constraints_compatible(
                         old_type_parameter.constraints,
                         new_type_parameter.constraints,
                     )
             },
         )
 }
 
+fn struct_layout_compatible(
+    old_struct: &StructDefinitionView<CompiledModule>,
+    new_struct: &StructDefinitionView<CompiledModule>,
+) -> bool {
+    if old_struct.variant_count() == 0 {
+        // Old is regular struct, new need to be as well (i.e. have zero variants) and compatible
+        // fields
+        new_struct.variant_count() == 0
+            && fields_compatible(
+                old_struct.fields_optional_variant(None),
+                new_struct.fields_optional_variant(None),
+            )
+    } else {
+        // Enum: the prefix of variants in the old definition must be the same as in the new one.
+        // (a) the variant names need to match
+        // (b) the variant fields need to be compatible
+        old_struct.variant_count() <= new_struct.variant_count()
+            && (0..old_struct.variant_count()).all(|i| {
+                let v_idx = i as VariantIndex;
+                old_struct.variant_name(v_idx) == new_struct.variant_name(v_idx)
+                    && fields_compatible(
+                        old_struct.fields_optional_variant(Some(v_idx)),
+                        new_struct.fields_optional_variant(Some(v_idx)),
```

Comment:
I assume there's a check somewhere else that the newly added variants are not the same as some of the ones already there?

Reply:
Its really we just checking that `old_variants prefix_of new_variants`, where two variants match if they have the same name and fields. So everything we need to check is here..